File: linked-list/MyLinkedList.java
public class MyLinkedList<T> {
  public Node<T> head;
  public Node<T> tail;
  public int size;

  public void addFirst(T value) {
    head = new Node<T>(value, head);
    size++;

    if (size == 1) {
      tail = head;
    }
  }

  public void add(T value) {
    if (tail == null) {
      head = tail = new Node<T>(value);
    } else {
      Node<T> newNode = new Node<T>(value);
      tail.next = newNode;
      tail = newNode;
    }
    size++;
  }

  public void add(T value, int index) {
    if (index < 0 || index > size) {
      throw new IndexOutOfBoundsException(index);
    }

    if (size == index) {
      add(value);
    } else if (index == 0) {
      addFirst(value);
    } else {
      Node<T> current = head;
      while (index != 1) {
        current = current.next;
        index--;
      }
      Node<T> temp = new Node<>(value, current.next);
      current.next = temp;
      size++;
    }
  }

  public void removeIndex(int index) {
    if (size == index) {
      throw new IndexOutOfBoundsException(index);
    } else if (index == 0) {
      head = head.next;
    } else {
      Node<T> current = head;

      while (index != 1) {
        current = current.next;
        index--;
      }

      current.next = current.next.next;
    }
    size--;
  }

  public boolean contains(T value) {
    Node<T> current = head;

    while (current != null) {
      if (current.value == value)
        return true;
      current = current.next;
    }

    return false;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    Node<T> current = head;

    if (current != null) {
      sb.append(current.value.toString());
      current = current.next;
    }

    while (current != null) {
      sb.append(" -> ");
      sb.append(current.value.toString());
      current = current.next;
    }

    return sb.toString();
  }

  public static void main(String[] args) {
    MyLinkedList<Integer> test = new MyLinkedList<Integer>();
    test.add(2);
    test.add(3);
    test.add(5);
    test.addFirst(1);
    test.add(4, 3);

    System.out.println(test.contains(4));
    System.out.println(test.toString());
    System.out.println("size" + " " + test.size);
    test.removeIndex(2);
    System.out.println(test.toString());
    System.out.println("size" + " " + test.size);
  }
}

class Node<T> {
  public T value;
  public Node<T> next;

  public Node(T value, Node<T> next) {
    this.value = value;
    this.next = next;
  }

  public Node(T value) {
    this.value = value;
    this.next = null;
  }
}

File: linked-list/MyLinkedList.ts
class MyNode<T> {
  public value: T;
  public next: MyNode<T> | null;

  constructor(value: T, next: MyNode<T> | null = null) {
    this.value = value;
    this.next = next;
  }
}

class MyLinkedList<T> {
  public head: MyNode<T> | null = null;
  public tail: MyNode<T> | null = null;
  public size: number = 0;

  public addFirst(value: T): void {
    this.head = new MyNode<T>(value, this.head);
    this.size++;

    if (this.size === 1) {
      this.tail = this.head;
    }
  }

  public add(value: T): void {
    if (this.tail === null) {
      this.head = this.tail = new MyNode<T>(value);
    } else {
      const newMyNode = new MyNode<T>(value);
      this.tail.next = newMyNode;
      this.tail = newMyNode;
    }
    this.size++;
  }

  public addAtIndex(value: T, index: number): void {
    if (index < 0 || index > this.size) {
      throw new Error(`Index out of bounds: ${index}`);
    }

    if (this.size === index) {
      this.add(value);
    } else if (index === 0) {
      this.addFirst(value);
    } else {
      let current: MyNode<T> | null = this.head;
      while (index !== 1 && current !== null) {
        current = current.next;
        index--;
      }
      if (current !== null) {
        const temp = new MyNode<T>(value, current.next);
        current.next = temp;
        this.size++;
      }
    }
  }

  public removeAtIndex(index: number): void {
    if (this.size === index) {
      throw new Error(`Index out of bounds: ${index}`);
    } else if (index === 0 && this.head !== null) {
      this.head = this.head.next;
    } else {
      let current: MyNode<T> | null = this.head;

      while (index !== 1 && current !== null) {
        current = current.next;
        index--;
      }

      if (current !== null && current.next !== null) {
        current.next = current.next.next;
        this.size--;
      }
    }
  }

  public contains(value: T): boolean {
    let current: MyNode<T> | null = this.head;

    while (current !== null) {
      if (current.value === value) {
        return true;
      }
      current = current.next;
    }

    return false;
  }

  public toString(): string {
    let sb: string = "";
    let current: MyNode<T> | null = this.head;

    if (current !== null) {
      sb += this.convertToString(current.value);
      current = current.next;
    }

    while (current !== null) {
      sb += ` -> ${this.convertToString(current.value)}`;
      current = current.next;
    }

    return sb;
  }

  public convertToString<T>(value: T): string {
    if (typeof value === "string") {
      return value;
    } else if (typeof value === "number") {
      return value.toString();
    } else {
      throw new Error("Cannot convert this type to string");
    }
  }
}

(() => {
  const test = new MyLinkedList<number>();
  test.add(2);
  test.add(3);
  test.add(5);
  test.addFirst(1);
  test.addAtIndex(4, 3);

  console.log(test.contains(4));
  console.log(test.toString());
  console.log("size", test.size);
  test.removeAtIndex(2);
  console.log(test.toString());
  console.log("size", test.size);
})();
